### DISCUSSION TITLE
Current status of AlarmServer

### DISCUSSION THREAD

**Discussion by rct** [juggie/AlarmServer#80] [General]
Here is an assessment of the current state of AlarmServer as of 2023-01-29:

## Why use AlarmServer in 2023?

Home automation platforms like Home Assistant with the PyEnvisalink-based integration have somewhat supplanted the need for AlarmServer's Web UI and REST/JSON interface. However:

* Currently there are problems with the robustness/reliability of the PyEnvisalink integration, see the discussion thread: [Home Assistant Community Form: Envisalink integration: Things users should know](https://community.home-assistant.io/t/envisalink-integration-things-users-should-know-about-it/326191/42).  The Envisalink integration does not appear to be actively maintained.
* The Envisalink only allows a single connection for TPI clients. If Home Assistant or other application connects directly to the Envisalink, no other software can connect to the Envisalink without shutting down Home Assistant. This makes developing and testing software difficult. Currently I don't believe there are any test harnesses / envisalink emulators that could be used for developing/testing without a live envisalink connected to an actual alarm panel.
* PyEnvisalink doesn't currently provide low-level access to the TPI interface. 

So current uses for AlarmServer include:
* Using AlarmServer's proxy functionality to enable multiple simultaneous connections to the Envisalink. Envisalink software like Home Assistant's envisalink integration can be connected to the AlarmServer proxy without modification(*). 
* Using AlarmServer to log the low-level TPI communications for monitoring, tracing, and debugging.
* Using AlarmServer for functionality not currently available in home automation integrations (zone bypass, toggling alarm panel programmable outputs, etc.)

### juggie's repo: 

* The master branch of [juggie/AlarmServer](https://github.com/juggie/AlarmServer) is the old code that was developed with Python2.7/Tornado 4.x.
* There is a Python 3 branch [feature/python3](https://github.com/juggie/AlarmServer/tree/feature/python3) that was dormant but has gotten new activity as of 2023-01-29 to work with Python 3 and a current version of Tornado (6.2.x)

### Other forks of AlarmServer

* [tinuva/AlarmServer](https://github.com/tinuva/AlarmServer) - Python 3, Tornado 6.1, as of 2023-01-29 @tinuva says he's using his running fork as his primary code, but doesn't consider his fork to be a replacement for Juggie's repo/branches.  This branch has also added bypass functionality to AlarmServer's web UI. 

**TBD: Are there other forks that should be listed here??**

## DSC vs Honeywell Ademco

The Envisalink supports both DSC and Honeywell panels, however the TPI interface is considerable different and incompatible. 

* Juggie/AlarmServer is only for use with DSC panels.
* There are forks AlarmServer for use with Honeywell Ademco panels -- [MattTW/HoneyAlarmServer](https://github.com/MattTW/HoneyAlarmServer) 

### PyEnvisalink Python Module

* [PyPi Index](https://pypi.org/project/pyenvisalink/)
* GitHub Repo: [ufodone/pyenvisalink](https://github.com/ufodone/pyenvisalink)  (Formerly this was Cinntax/pyenvisalink on GitHub, but was transferred in 2022 from @cinntax to @ufodone. 

From [PyEnvisalink's README.md](https://github.com/ufodone/pyenvisalink/blob/master/README.md)

> This project is an attempt at creating a pure, raw python3 library for interfacing with the envisalink line of alarm products. In particular the envisalink EVL3 and EVL4. This project leverages the asyncio library, providing full asynchronous communication with the envisalink board. All that must be done is call the proper constructor with connectivity/credentials, and provide callbacks for any events you wish to be notified of.

> Also, an attempt will be made at merging the two APIs that are available for the Envisalink alarm interfaces (DSC and Honeywell). While the APIs are quite a bit different, a base functionality is common between them, and thus an abstraction can be made. We'll see how that works in practice ;).

> This project was originally a fork of the [Envisalink 2DS/3 Alarm Server for Honeywell or Ademco Vista Security Systems](https://github.com/MattTW/HoneyAlarmServer) - Matt's focus was much more of an "all-in-one" approach, where my aim is to just provide a python API that can be used within something else. While our codebases are now a bit separate due to my use of a different asynchronous library, his examples and logic were invaluable!

### PyEnvisalink NEW updated Python Module

* [ufodone/envisalink_new](https://github.com/ufodone/envisalink_new)

New in 2023-01, @ufodone is working on a rewrite of PyEnvisalink:

From the readme 

> Current changes include:
>
> * Full support for UI configuration flow (configuration.yaml to be deprecated)
> * Support for multiple envisalink devices
> * Entities now have unique IDs allowing configuration/customization via the UI
> * Zone bypass switch support for DSC panels
> * Auto-discovery of EVL version and panel type (DSC/HONEYWELL)
> * Refactoring of the underlying pyenvisalink package including
>    * Sequential queueing of commands to the EVL including retry on errors (which applicable) and timeouts
>    * Ability to query EVL firmware version and MAC address
>    * Update of asyncio network handling to use Streams rather than low-level APIs

### PyEnvisalink and AlarmServer

There is some overlap between the PyEnvisalink Python module and the DSC and Honeywell Ademco AlarmServer implementations.  Ideally AlarmServer could use PyEnvisalink to communicate with the EnvisaLink.  @Juggie has begun looking at that. His initial take is that PyEnvisalink may not provide enough level access to TPI communications to enable PyEnvisalink to be used as part of a proxy to the Envisalink.

### Making the Home Assistant Envisalink integration more reliable/robust.

TBD - most likely this should be a separate discussion

### MQTT interface

@tinuva had the idea of adding an MQTT interface to AlarmServer,. This would be similar to ZwaveJS2MQTT, Zigbee2MQTT, and [Paradox Alarm Interface (PAI)](https://github.com/ParadoxAlarmInterface/pai). 

Home Assistant has an [MQTT Alarm integration](https://www.home-assistant.io/integrations/alarm_control_panel.mqtt/) that allows alarm systems to expose a somewhat generic MQTT interface that could then be used by Home Assistant (and presumably other platforms) without having to develop a custom integration for each one.

### Merging DSC and Honeywell Ademco code bases 

PyEnvisalink is endeavoring to provide a single library to talk to either family of envisalink connected alarm panels.  This should most likely be a larger/separate discussion. However I think it is worth considering there are at least two distinct components here:

1. The communications framework and mechanisms that aren't specific to either alarm panel family
2. The alarm panel specific modeling of state for the panel, zones, partitions, etc. 

Ideally Envisalink clients for either panel family could share No. 1 to reduce duplication of effort. PyEnvisalink and the move to Python Asyncio may make this possible. A generic envisalink proxy that enables multiple connections if it could be implemented and useful without having to implement too much panel state tracking would be a benefit that users could deploy that between their envisalinks and their chosen software.

In terms of an abstraction for modeling generic alarm panel's, it is probably work looking at:
* [Home Assistant's MQTT Alarm integration](https://www.home-assistant.io/integrations/alarm_control_panel.mqtt/) 
* [nielsfaber/alarmo](https://github.com/nielsfaber/alarmo) - my take on this project is that it creates a virtual alarm panel in Home Assistant that allows users to cobble together their own alarm system from existing entities (such as Zigbee/Zwave/Matter/ESP32 etc sensors) in Home Assistant. Efforts like this, in my opinion, are likely to become increasingly relevant--there is an ever growing amount  of cost-attractive devices to provide home security and monitoring functionality from multiple vendors across different communication domains.

**Comment by juggie** [comment]
Very good writeup.  I've started wrapping my head around asyncio (I have only really used tornado!).  And time permitting I should have pyenvisalink ported from using the protocol interface to the stream interface shortly(ish).  That should help with the robustness of the parsing which from a quick glance has a bug due to how data is read off the buffer.

**Comment by tinuva** [comment]
@ufodone I believe also started work on https://github.com/ufodone/envisalink_new which does have the bypass switches.
I had some interesting [issue](https://github.com/ufodone/envisalink_new/issues/2) to get it to work over alarmserver, as I also have to run a reverse proxy for the web interface to make that version work.

**Reply by rct** [reply]
@tinuva thanks! - I completely missed that, particularly given that he is using a new repo instead of a branch.   So if I read through that issue will I figure out how to get it running connected to AlarmServer?

**Reply by ufodone** [reply]
My plan is to move into into the Home Assistant repo but decide to start with a HACS version because it's easier to iterate on and get others to help test.  This repo will eventually go away once it's added to HA.  The pyenvisalink copy within the envisalink_new repo is, again, a convenience to avoid needing to publish versions of that library while I'm iterating on things.  However, I am keeping those changes up to date in a separate branch in the `pyenvisalink` repo.

**Reply by tinuva** [reply]
> @tinuva thanks! - I completely missed that, particularly given that he is using a new repo instead of a branch. So if I read through that issue will I figure out how to get it running connected to AlarmServer?

No complete example in there. You just have to run a reverse proxy like nginx or traefik or your preference, on the same ip address which you run alarmserver. It was port 80 but now it is port 8080 on which you need to run it. His repo is still changing often as it is in dev/testing phase now.

**Comment by ufodone** [comment]
@juggie I've got a version of pyenvisalink working using streams.  It's part of a larger overhaul of the library.  You can find it [here](https://github.com/ufodone/pyenvisalink/tree/command_queuing).

I encountered the parsing bug you refer to as part of doing that switch and have fixed it (although it needs some cleanup/simplification because at the time I didn't have enough data on Honeywell systems to go on to know what those units sent).

**Comment by juggie** [comment]
@ufodone oh nice that will save me some time.

I'll have a look at that.  One thing I'll implement if it's not there is raw access to the events so they can be relayed if so desired (as my proxy functionality does).

I'll have a look and let you know what I'm thinking.

**Comment by juggie** [comment]
@ufodone Unless i'm mistaken, the implementation here: https://github.com/ufodone/pyenvisalink/blob/command_queuing/pyenvisalink/envisalink_base_client.py is still using asyncio.Protocol, my goal would be to eliminate the class extending (it's not recomended) and then use the asyncio Streams which then gets you access to readuntil().  So we can then properly read until the command seperator that the envisalink module does.  The way its currently implemented will risk getting a partial command, and there's nothing here: https://github.com/ufodone/pyenvisalink/blob/command_queuing/pyenvisalink/envisalink_base_client.py#L277 to handle left over data and keep it for the next time you get an async io callback.  If I missed something or maybe you just didn't push some code let me know!

**Reply by rct** [reply]
@juggie - @ufodone is working in a separate repo - https://github.com/ufodone/envisalink_new  (note the _new) and the other comment, that he's currently working on a copy of pyenvisalink that is inside of the `_new` repo to make it easier for people to test via HACS.

**Reply by ufodone** [reply]
Ah, that's just a bit of leftover I need to cleanup.  While the class still extended asyncio.Protocol, it doesn't actually use it (I just forgot to remove it). It doesn't actually do anything anymore.  The `process_data()` function is caledl from the read_loop() which is using Streams.  The leftover data is handled in read_loop as well.

**Reply by juggie** [reply]
Best way you do it would be to use reader.readuntil().  Then you are guaranteed to get one line at a time, no need for iteration, no need to split, etc.

**Reply by ufodone** [reply]
At the time I wrote it, I had no info about the Honeywell systems other than the TPI documentation which doesn't actually indicate that there are CR/LFs between the commands so I couldn't rely on readuntil().  Like I said, I need to go back and clean it up now that I have more info available as to what the envisalink devices actually send.

**Reply by juggie** [reply]
> At the time I wrote it, I had no info about the Honeywell systems other than the TPI documentation which doesn't actually indicate that there are CR/LFs between the commands so I couldn't rely on readuntil().  Like I said, I need to go back and clean it up now that I have more info available as to what the envisalink devices actually send.

Ok sounds good.  I'll flip to that branch and have a look.

**Comment by juggie** [comment]
@ufodone Did you have a plan for how to run your branch of pyenvisalink ie in testing.  This is what I got working:
```python
import asyncio
from pyenvisalink import EnvisalinkAlarmPanel

loop = asyncio.get_event_loop()

panel = EnvisalinkAlarmPanel('envisalink', eventLoop=loop)

async def main():
    x = await panel.discover()
    x = await panel.start()

loop.create_task(main())
loop.run_forever()
```

**Reply by juggie** [reply]
EnvisalinkAlarmPanel doesn't seem to want to make it's own event loop anymore with the creation of discover/start.

Was 'discover' created to attempt to get Home Assistant to discover the device on the network?

**Reply by juggie** [reply]
One more note, I forked pyenvisalink and created a branch here: https://github.com/juggie/pyenvisalink/tree/feature/code-cleanup and have been doing a bit of cleanup ie flake8, black, etc.  There's some issues left still, some code that won't compile (but doesn't get hit) and so on.

**Reply by ufodone** [reply]
Yeah, I keep forgetting to update the test_harness.py.  I have my own copy that I've been using but haven't consolidated it into test_harness.py.  What you posted is pretty similar to what I have been using.

I added `discover()` as a way to query the device for as much info as I could get and to validate that credentials were valid.  I use this in the Home Assistant integration for the config flow setup and validation.  I also didn't want to have to start the loop during configuration hence breaking it out as a separate function.

It's still honestly a bit of a work in progress as I'm still settling on how I'd like things to work.  Had a bit of a discussion on it over [here](https://github.com/ufodone/envisalink_new/issues/10).

There's still code in there to start its own event loop but I'll admit I haven't tested with it for quite a while.  I'm certainly no expert on this stuff but wasn't sure whether supporting creating its own event loop helped that much beyond saving a couple lines of code on initial setup.

Regarding the code cleanup: happy to take any of those changes back.  To date I've mostly been focused on getting things working but know I'll need to go back and clean stuff up at some point.

**Reply by juggie** [reply]
I actually just removed the requirement to pass event loop in as it wasn't needed.  Async io is already able to discover an existing asyncio and you can't run more than one at a time.  All that code is pushed up.  I also just implemented the readuntil().

I also don't think the start() making it's own ioloop and discover() work together as discover is an async function so the loop needs to exist to use that.  And it needs to be called before start().  I was also going to look at having start call it automatically.

**Reply by ufodone** [reply]
I also made the `readuntil()` changes earlier today... maybe after you forked?

**Reply by ufodone** [reply]
I also originally had start() call discover() but decided to separate it out.  At the moment I'm leaning towards trying to have start() not depend on discover() at all.  It currently needs the EVL version and panel type but I think I can infer those automatically so discover() could be used independently (and likely just for HA setup).

**Reply by juggie** [reply]
> I also originally had start() call discover() but decided to separate it out.  At the moment I'm leaning towards trying to have start() not depend on discover() at all.  It currently needs the EVL version and panel type but I think I can infer those automatically so discover() could be used independently (and likely just for HA setup).
> 

That would likely make sense there's no reason to force it if users provide it.

**Comment by juggie** [comment]
@ufodone I re-branched and re-applied all my work (some of it I re-ran the tools as the merge was a pain).  It's just an initial tidy-up, get flake8 less angry, etc.  I have a few ideas I want to implement so that I can implement pyenvisalink within AlarmServer.

I opened a PR.

**Comment by mathompl** [comment]
I have been having constant problems with hanging connections to my EVL3 - sometimes once a day. Only proxy or evl restart fixes this.

It seems that evl3 randomly stops responding but keeps opened connection and alarm server cannot reconnect properly - it never closes connection, so evl doens't let it reconnect (one tpi client limit)

I think I've tried every single proxy for evl out there. All have problems, this one is imho the best, most reliable one with exception to the above problem.  

I've decided to try to fix the code, here is the result (with docker). It also fixes tedious login process and adds keepalive ping to evl.
https://github.com/mathompl/AlarmServer-fixed

in the process i migrated it to py3.8, cleaned up the code, run scripts, exceptions, logging, created docker and fixed http/https authorization.

I'm using it with 2-3 clients without problems (home assistant and my java client)

**Reply by mathompl** [reply]
btw: it also works with Home Assistant's envisalink_new integration now.
